### PR TITLE
Next step unlocked persistence

### LINKED ISSUE #9952: Slide challenges: navigating back locks previously unlocked button
#### Challenge Name

Any slide challenge with an "open link in new tab" button. Here's one: http://beta.freecodecamp.com/en/challenges/join-the-free-code-camp-community/create-a-github-account-and-join-our-chat-rooms
#### Issue Description

If I click the "open link in new tab" button, then click the "go to my next step" move to the next step, then navigate back, the "go to my next step" button will be locked again
#### Browser Information

- Browser Name, Version: 
- Operating System: 
- Mobile, Desktop, or Tablet: 
#### Your Code

``` js

```

### PR DESCRIPTION
**Pull request by Bouncey**
#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of FreeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [x] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [x] All new and existing tests pass the command `npm run test-challenges`. Use `git commit --amend` to amend any fixes.
#### Type of Change

- [x] Small bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)
#### Checklist:

- [x] Tested changes locally.
- [x] Closes currently open issue (replace XXXX with an issue no): Closes #9952
#### Description

Utilises localStorage to persist next step unlocked.

### CODE REVIEW

File: common/app/routes/challenges/redux/actions.js [hunk 1]
```diff
@@ -1,14 +1,17 @@
...
+import { getMouse, loggerToStr, unlockStep } from '../utils';
 
 import types from './types';
 
 // step
 export const stepForward = createAction(types.stepForward);
 export const stepBackward = createAction(types.stepBackward);
 export const goToStep = createAction(types.goToStep);
-export const completeAction = createAction(types.completeAction);
+export const completeAction = createAction(
```
Comment (BerkeleyTrue) [inline common/app/routes/challenges/redux/actions.js]: Action creators should not have side effects.

File: common/app/routes/challenges/utils.js [hunk 2]
```diff
@@ -488,3 +488,34 @@ export function unfilterMapUi(tree) {
     }
   );
 }
+
+export function unlockStep(challengeId, index) {
+  if (!challengeId || !index) {
+    return false;
+  }
+  if (typeof Storage !== 'undefined') {
```
Comment (BerkeleyTrue) [inline common/app/routes/challenges/utils.js]: These utils should not have side effects. 

Side effects should happen in an epic.

File: common/app/routes/challenges/components/step/Step.jsx [hunk 3]
```diff
@@ -79,22 +82,25 @@ export class StepChallenge extends PureComponent {
...
-  renderActionButton(action, completeAction) {
+  renderActionButton(id, index, action, completeAction) {
     const isApiAction = action === '#';
     const buttonCopy = isApiAction ?
       'Confirm' :
       'Open link in new tab ';
     if (!action) {
       return null;
     }
+    function onClickHandler() {
```
Comment (BerkeleyTrue) [inline common/app/routes/challenges/components/step/Step.jsx]: Functions should not be created per render. Use an instance method

**Review by BerkeleyTrue — CHANGES_REQUESTED** [review]

**Comment by Bouncey** [comment]
I will update the tests to cover for the new payload from `step-challenge-epic`

**Comment by BerkeleyTrue** [comment]
We shouldn't put these in localStorage. There is no reason too.

The original issue is with going back while in a step challenge. We can just put this information in the redux store.

Something like an array of booleans indicating if the step action has been taken.

File: common/app/routes/challenges/redux/step-challenge-epic.test.js [hunk 4]
```diff
@@ -56,7 +62,7 @@ test(file, function(t) {
           t.assert(
             onNextSpy.calledWithMatch({
               type: types.goToStep,
-              payload: 0
+              payload: { step: 0, isUnlocked: false }
```
Comment (Bouncey) [inline common/app/routes/challenges/redux/step-challenge-epic.test.js]: I don't know how to set this test up to return true.

Guidance would be greatly appreciated :+1:

File: common/app/routes/challenges/components/step/Step.jsx [hunk 5]
```diff
@@ -79,7 +81,12 @@ export class StepChallenge extends PureComponent {
     }
   }
 
-  renderActionButton(action, completeAction) {
+  handleActionClick() {
```
Comment (Bouncey) [inline common/app/routes/challenges/components/step/Step.jsx]: This needs removed, left over from a previous iteration

**Comment by Bouncey** [comment]
I have updated the tests to check that the previous unlocked step remains unlocked on `stepBackward`

This is ready for another QA @BerkeleyTrue

File: common/app/routes/challenges/redux/step-challenge-epic.js [hunk 6]
```diff
@@ -1,26 +1,59 @@
 import types from './types';
-import { goToStep, submitChallenge } from './actions';
+import { goToStep, submitChallenge, updateUnlockedSteps } from './actions';
 import { challengeSelector } from './selectors';
 import getActionsOfType from '../../../../utils/get-actions-of-type';
 
+function unlockStep(challengeId, index, unlockedSteps) {
```
Comment (BerkeleyTrue) [inline common/app/routes/challenges/redux/step-challenge-epic.js]: We don't want to save all the challenge steps completed just the current one.

File: common/app/routes/challenges/redux/step-challenge-epic.js [hunk 7]
```diff
@@ -1,26 +1,59 @@
...
+  updated = unlockedSteps[challengeId] = [index];
+  updateUnlockedSteps(updated);
+  return null;
+}
+
+function isStepUnlocked(challengeId, index, unlockedSteps) {
+  let current = unlockedSteps[challengeId];
+  if (current) {
+    return (current.some( i => i === index));
+    }
```
Comment (BerkeleyTrue) [inline common/app/routes/challenges/redux/step-challenge-epic.js]: indenting

File: common/app/routes/challenges/redux/step-challenge-epic.js [hunk 8]
```diff
@@ -1,26 +1,59 @@
...
+  }
+  updated = unlockedSteps[challengeId] = [index];
+  updateUnlockedSteps(updated);
+  return null;
+}
+
+function isStepUnlocked(challengeId, index, unlockedSteps) {
+  let current = unlockedSteps[challengeId];
+  if (current) {
+    return (current.some( i => i === index));
```
Comment (BerkeleyTrue) [inline common/app/routes/challenges/redux/step-challenge-epic.js]: padding in the arguments
unnecessary parenthesis around expression.

File: common/app/routes/challenges/redux/step-challenge-epic.js [hunk 9]
```diff
@@ -1,26 +1,59 @@
...
 import getActionsOfType from '../../../../utils/get-actions-of-type';
 
+function unlockStep(challengeId, index, unlockedSteps) {
+  if (!challengeId || !index) {
+    return null;
+  }
+  let current = unlockedSteps[challengeId];
+  let updated;
+  if (current) {
+    updated = unlockedSteps[challengeId] = current.concat([index]);
```
Comment (BerkeleyTrue) [inline common/app/routes/challenges/redux/step-challenge-epic.js]: You are mutating the state directly. You should never mutate state. State should be updated through the reducer

**Comment by BerkeleyTrue** [comment]
While the code quality looks good I'm concerned about the way locked steps are stored. They are currently stored in a hash `unlockedSteps: { [challengeId]: [...Number] }`. 

This poses a couple of problems. 

We are storing information in state of challenges between subsequent challenges. This works fine until the user refreshes the page. They will start out with an empty state, causing confusion when a previously completed action from another challenge is no longer completed. 

The second issue is the storage mechanism itself is inefficient. Steps numbers are stored in an array which is stored in a hashmap keyed to each challenge id. The properties of the array data structure itself could be used instead of storing the information of which step is completed by pushing the step number onto an array. 

`unlockedSteps: [...Boolean|Void]`

To check if a step has been completed we simply do `!!unlockedSteps[stepIndex]`;

This should solve both of the issues above and doing an index check is more efficient then iterating through all those elements.

**Comment by Bouncey** [comment]
The reason I put it in a hash was to cut down possible confusion if the camper jumps around the challenges and sees steps unlocked when they haven't completed the required action.

If you need `stepsUnlocked` to persist over a refresh, I can place them into local storage with a method to remove them when the challenge is submitted. 

Your idea of an array of steps would require more iteration than my hash. It would need to hold a bool for each step in the challenge, compared to my hash that only stores a step if required.

**Comment by BerkeleyTrue** [comment]
> if the camper jumps around the challenges and sees steps unlocked when they haven't completed the required action.

These steps should always be reset on challenge load. 

> If you need stepsUnlocked to persist over a refresh

No they shouldn't.

> Your idea of an array of steps would require more iteration than my hash

Pushing numbers onto an array requires iteration. Using the index of the array would require no iteration.

You don't need to hold a bool for each step, just the completed steps. You can assume `undefined` is uncompleted or no action required.

**Review by BerkeleyTrue — APPROVED** [review]

**Comment by QuincyLarson** [comment]
@Bouncey thanks for fixing this!

**Pull request merged**